### PR TITLE
Ground truth

### PR DESCRIPTION
**Pull request by neworderofjamie**
After some thought, one of the issues with the loss function implementations was that they mixed up the concepts of calculating losses with providing ground truth information. This was a cause of mess for several reasons:

1. Multiple loss functions use the same type of ground truth (``RelativeMeanSquareError`` and ``SparseCategoricalCrossentropy`` both require standard per-example labels)
2. The handling of ground truth was mixed up with everything else in the ``_build_out_neuron_model`` mess.
3. In order to correctly implement metrics like MSE, these bits of code will also need access to the ground truth - this definitely shouldn't be mixed up with loss functions as this will also be required in inference compiler

This PR tidies this up by introducing ``GroundTruth`` classes. Loss function classes are stripped back to just providing a property which specifies which ground truth they require and these are created using the normal mlGeNN mechanism. These ground truth classes then deal with everything to do with ``YTrue`` and ``YBack`` with methods called at the start and end of ``_build_out_neuron_model``.

### CODE REVIEW

File: ml_genn/ml_genn/compilers/ground_truths/example_value.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    def push_to_device(self, genn_pop, y_true, shape, batch_size: int,
+                       example_timesteps: int):
+        # Check shape
+        expected_shape = (batch_size,) + shape
+        y_true = np.asarray(y_true)
+        if y_true.shape != expected_shape:
+            raise RuntimeError(f"Shape of target data for ExampleValue "
+                               f"ground truth should be {expected_shape}")
+
+        # Copy flattened y_true into view
```
Comment (tnowotny) [inline ml_genn/ml_genn/compilers/ground_truths/example_value.py]: y_true doesn't seem to be flattened; why is the view 2D - is this correct?

**Review by tnowotny — COMMENTED** [review]
I was surprised at first that losses disappeared from compiled training network, but I see that they are handled in the compiler and then don't necessarily appear anymore. However, will there be a future provision that users can read out loss values besides accuracy values? Might need to add loss information to the compiled network again at that point? Also see comment.

File: ml_genn/ml_genn/compilers/ground_truths/example_value.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply ml_genn/ml_genn/compilers/ground_truths/example_value.py]: So I added some unit tests to test this logic and, actually, the batch 1 logic was fine as ``y_true`` always has a batch dimension **but** ``ExampleValue`` and ``TimestepValue`` did not correctly handle the end of the batch
